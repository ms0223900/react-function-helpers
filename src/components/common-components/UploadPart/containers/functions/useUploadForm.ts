import useDragDropFiles from "../../lib/useDragDropFiles";
import useUploadAndPreviewFiles, { UseUploadAndPreviewFilesOptions } from "../../lib/useUploadAndPreviewFiles";


// 自訂訊息
const onUploadSuccess = (e: any) => {
  window.alert("Upload Success!");
};
const onUploadFailed = (e: any) => {
  window.alert(`Upload failed: ${e.message}`);
};

export interface UseUploadFormOptions extends Pick<UseUploadAndPreviewFilesOptions, 'uploadApiUrl' | 'maxUploadFileAmount'> {
}

const useUploadForm = ({
  uploadApiUrl, maxUploadFileAmount=Infinity
}: UseUploadFormOptions) => {
  const uploadAndPreviewFilesState = useUploadAndPreviewFiles({
    uploadApiUrl,
    maxUploadFileAmount,
    onUploadSuccess,
    onUploadFailed
  });

  const dragDropFilesState = useDragDropFiles({
    onSetDropFiles: uploadAndPreviewFilesState.handleAddFiles
  });

  return {
    ...uploadAndPreviewFilesState,
    ...dragDropFilesState
  };
};

export default useUploadForm;
