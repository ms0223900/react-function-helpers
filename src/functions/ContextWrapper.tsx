import { Callback } from 'all-common-types';
import React, { Context, ReactNode } from 'react';

export interface ContextWrapperProps<ContextState> {
  customInitState?: Partial<ContextState>
  customDispatch?: Callback
  children: ReactNode
}
export type ContextValueFnType<ContextState> = (
  customInitState?: Partial<ContextState>,
  customDispatch?: Callback,
) => {
  dispatch: (x?: any) => any
  state: ContextState
}

export function ContextWrapperFn<ContextState>(
  ContextValueFn: ContextValueFnType<ContextState>,
  ContextStore: Context<{
    dispatch: (x?: any) => any
    state: ContextState
  }>
) {
  return (props: ContextWrapperProps<ContextState>) => {
    const { children, customInitState, customDispatch } = props;
    const value = ContextValueFn(customInitState, customDispatch);
    return (
      <ContextStore.Provider value={value}>
        {children}
      </ContextStore.Provider>
    );
  };
}

export default ContextWrapperFn;