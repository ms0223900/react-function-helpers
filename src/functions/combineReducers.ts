type Reducers<State> = {
  [name in keyof State]: (state: State, action: any) => State[name]
}
function combineReducers<State extends object>(reducers: Reducers<State>) {
  type Key = keyof State
  return (state: State, action: any): State => {
    let newState = state;
    const allReducers = Object.keys(reducers) as Key[];
    for (const reducer of allReducers) {
      newState = {
        ...newState,
        [reducer]: reducers[reducer](newState, action),
      };
    }
    return newState;
  };
}

export default combineReducers;