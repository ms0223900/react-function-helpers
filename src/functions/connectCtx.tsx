import React, { Dispatch, useContext } from 'react';
import { ComponentType, Context } from "react";
import makePropsFromContext, { MapDispatchToProps, MapStateToProps } from './mapContextToProps';

export type ContextStoreGeneric<State> = Context<{
  state: State
  dispatch: Dispatch<any>
}>

function connectCtx<State extends object>(contextStore: ContextStoreGeneric<State>) {
  return function mapCtxToProps<OwnProps, PropsFromState, PropsFromDispatch>(
    mapStateToProps?: MapStateToProps<State, OwnProps, PropsFromState>,
    mapDispatchToProps?: MapDispatchToProps<OwnProps, PropsFromDispatch>
  ) {
    type Props = OwnProps & PropsFromState & PropsFromDispatch
    return (Component: ComponentType<Props>) => {
      return (props: OwnProps) => {
        const { state, dispatch } = useContext(contextStore);
        const propsFromState = makePropsFromContext(state, props, mapStateToProps);
        const propsFromDispatch = makePropsFromContext(dispatch, props, mapDispatchToProps);
        return (
          <Component 
            {...props} 
            {...propsFromState}
            {...propsFromDispatch} />
        );
      };
    };
  };
}

export default connectCtx;