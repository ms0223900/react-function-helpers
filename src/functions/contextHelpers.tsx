import { Obj } from "all-common-types";
import { createContext, useReducer } from "react";

type Dispatch = (x?: any) => any
type Reducers<State> = (state: State, action: any) => State
export interface MyContext<State> {
  dispatch: Dispatch
  state: State
}

export function createContextStore<State extends Obj>(
  initState: State, 
  dispatch: Dispatch=() => {
    // dispatch
  }
) {
  return createContext<MyContext<State>>({
    dispatch,
    state: initState
  });
}

export function createContextValueFn<State extends Obj>(
  initState: State, 
  reducers: Reducers<State>
) {
  return (customInitState?: Partial<State>, customDispatch?: Dispatch) => {
    const [state, dispatch] = useReducer(reducers, {
      ...initState,
      ...customInitState
    });
    return ({
      dispatch: customDispatch ? customDispatch : dispatch,
      state
    });
  };
}

export default {
  createContextStore,
  createContextValueFn
};