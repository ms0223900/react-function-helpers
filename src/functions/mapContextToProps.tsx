import { Dispatch } from "react";

type StateObj = object
export type StateOrDispatch<Context extends Dispatch<any> | StateObj> = Context extends Dispatch<any> ? Dispatch<any> : Context

export type MapContextToProps<Context extends Dispatch<any> | StateObj, OwnProps, PropsFromContext> = (
  stateOrDispatch: Context,
  ownProps: OwnProps,
) => PropsFromContext

export type MapDispatchToProps<OwnProps, PropsFromDispatch> = MapContextToProps<Dispatch<any>, OwnProps, PropsFromDispatch>
export type MapStateToProps<State extends object, OwnProps, PropsFromState> = MapContextToProps<State, OwnProps, PropsFromState>

export function makePropsFromContext<State extends StateObj, C extends Dispatch<any> | State, O, P>(
  context: C,
  ownProps: O,
  mapContextToProps?: MapContextToProps<C, O, P>
) {
  let propsForDispatch = {} as P;
  if(mapContextToProps) {
    propsForDispatch = mapContextToProps(context, ownProps);
  }
  return propsForDispatch;
}

export default makePropsFromContext;
