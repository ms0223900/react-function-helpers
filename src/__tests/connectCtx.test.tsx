import { mount, shallow } from 'enzyme';
import connectCtx from 'functions/connectCtx';
import { MapContextToProps, MapStateToProps } from 'functions/mapContextToProps';
import React, { createContext, Dispatch } from 'react';
import renderer, { act } from 'react-test-renderer';

describe('test connectCtx function', () => {
  const context = {
    dispatch: (x: any) =>x,
    state: {
      a: 'a',
      b: 'b'
    },
  };
  type State = typeof context['state']
  const contextStore = createContext(context);
  const Component = () => <></>;

  it('should wrap component and render correctly', () => {
    const ConnectCtxComponent = connectCtx(contextStore)()(Component);
    const tree = renderer.create(
      <ConnectCtxComponent />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should map state to props if parameter has mapStateToProps', () => {
    const mapStateToProps: MapStateToProps<State, {}, { a: string }> = (state) => {
      return ({
        a: state.a,
      });
    };
    const MyComponent = (props: {
      a: string
    }) => <>{props.a}</>;
    const ConnectCtxComponent = connectCtx(contextStore)(mapStateToProps)(MyComponent);
    const wrapper = mount(
      <ConnectCtxComponent />
    );
    expect(wrapper.find(MyComponent).props().a).toEqual(context.state.a);
  });

  it('should map dispatch to props if parameter has mapDispatchToProps', () => {
    const fn = jest.fn().mockImplementation((x: string) => x);
    const mapDispatchToProps: MapContextToProps<Dispatch<any>, {
      a: string
    }, {
      fn: (x: string) => any
    }> = (dispatch) => {
      return ({
        fn: (x) => dispatch(fn(x)),
      });
    };
    const MyComponent = (props: {
      fn: (x: string) => any,
      a: string
    }) => <></>;

    const ConnectCtxComponent = connectCtx(contextStore)(undefined, mapDispatchToProps)(MyComponent);
    const wrapper = mount(
      <ConnectCtxComponent a={''} />
    );
    act(() => {
      wrapper.find(MyComponent).props().fn('hello');
    });
    expect(fn).toBeCalledWith('hello');
  });

});