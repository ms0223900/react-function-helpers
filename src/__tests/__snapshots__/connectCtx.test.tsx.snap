// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test connectCtx function should wrap component and render correctly 1`] = `null`;
